philosophy/FORMS.md

```diff
@@ -0,0 +1,928 @@
+
+Sometimes we want to build input elements which deal with multiple values at once.
+One such element is our [`MoneyInput`](https://uikit.commercetools.com/?selectedKind=Inputs&selectedStory=MoneyInput). It allows to set a currency and an amount by using a dropdown for the currency and an input for the amount.
+
+The value produced by `MoneyInput` is represented as `{ currencyCode: 'EUR', amount: '2.00' }`
+
+> Sidenote: As the `MoneyInput` uses monetary formatting, we are using a text-input. This allows us to format values properly (e.g. by adding the trailing ".00") without browsers messing with the input. It also allows us to use decimal separators, e.g. for "1,000.00".
+
+We can construct our Formik value like this:
+
+```js
+{
+  id: 'party-parrot',
+  version: 1,
+  name: 'Party Parrot',
+  price: { currencyCode: 'EUR', amount: '2.00' }
+}
+```
+
+How do we now build a single input which receives `{ currencyCode: 'EUR', amount: '2.00' }` as its value and is able to change both using Formik's `handleChange`? We want to be able to render `<MoneyInput onChange={handleChange} onBlur={handleBlur} />` to make it as straightforward as using `<input type="text" />`.
+
+There are two approaches:
+
+- We can ensure `MoneyInput` always calls `onChange` with a full value like `{ currencyCode: 'EUR', amount: '2.00' }`
+- We can adapt the input names to be `price.currencyCode` or `price.amount` respectively. Formik will read that name and set the nested value only.
+
+If we were to go with the first approach, we'd need to fake `event.target.name` in `onBlur` as well, otherwise formik would set `{ touched: { price: true } }` instead of `{ touched: { price: { amount: true, currencyCode: true } } }`. That is bad, as Formik will set ``{ touched: { price: { amount: true, currencyCode: true } } }`once the form is submitted - so we end up with a`touched`state which changes shape! There is a CodeSandbox which illustrates this problem. Try changing the price and keep an eye on the`touched`state. Then submit the form, and you'll see a different`touched` state.
+
+If we were to accept changes to the shape of our `touched` state, we'd need to tak the different shapes into consideration every time we work with the `touched` shape and it just generally makes things harder to follow.
+
+So it's better to go with the second approach right away. We amend the names of the inputs by adding the `.amount` and `.currencyCode` suffix to their names and ids. When a parent doesn't pass in a name/id, we omit the property from the underlying input. This has solved that problem nicely for us.
+
+### Form values need to be shown in a user-friendly format
+
+As mentioned above, our `MoneyInput` consists of two parts: an currency selector and an amount input.
+
+We want the amount to be formatted like a monetary value, e.g. "200.00" or "2,000.00". However, the data we receive from our fantasy API looks like this: `{ currencyCode: 'EUR', centAmount: 200 }` for 2€.
```

Comment:
```suggestion
We want the amount to be formatted like a monetary value, e.g. "200.00" or "2,000.00". However, the data we receive from our API looks like this: `{ currencyCode: 'EUR', centAmount: 200 }`, which means **2€**.
```

Reply:
"Fantasy" just sounds weird in this context. A better word might be "fictitious" if you want to point out that it's not a real API.

Reply:
Yeah that's better! Done 👍

Reply:
Yeah I think the ~best~ english term would be hypothetical API 😆 :trollface: